### PR TITLE
Add sincpi, coscpi, sincu, coscu, and kwargs to sinc, cosc

### PR DESCRIPTION
**Pull request by moble**
The commenters on [this discourse post](https://discourse.julialang.org/t/unnormalized-sinc/25015) and [this github issue](https://github.com/JuliaLang/julia/issues/36954#issuecomment-833020887) seemed to agree that some thing needed to be done about the `sinc` and `cosc` functions related to the choices of normalization.  Specifically, I think this sums up the suggestions in those threads:

  1. Add a keyword argument `normalized=true` to `sinc`.
  2. Introduce `sincpi` to return the normalized `sinc` function, and to rhyme with `sincpi`, `cospi`, etc.
  3. Introduce `sincu` to return the unnormalized `sinc` function.
  4. Add a keyword argument `normalized=true` to `cosc`
  5. Introduce `coscpi`
  6. Introduce `coscu`
  7. Add documentation suggesting that the explicit forms should be used, to pave the way for deprecation in the future. 

This PR does all of those *except* the last one, as I'm not sure how likely it is that such a thing will ever happen.  I welcome input on that, or any other issue.

@stevengj @PallHaraldsson @KronosTheLate

### CODE REVIEW

File: base/special/trig.jl [hunk 1]
```diff
@@ -1105,19 +1160,93 @@ function _cosc(x::Number)
...
+    # cancellation error near x=0, so we use the Taylor series
+    # for small enough |x|.
+    if fastabs(x) < 1.57
+        # generic Taylor series: ∑ (-1)^n (x)^{2n-1}/a(n) where
+        # a(n) = (1+2n)*(2n-1)! (= OEIS A174549)
+        s = (term = -x)/3
+        x² = term^2
+        ε = eps(fastabs(term)) # error threshold to stop sum
+        n = 1
+        while true
```
Comment (oscardssmith) [inline base/special/trig.jl:1229]: Is this really the best implimentation?

Comment (moble) [reply base/special/trig.jl:1229]: Well, given that there are explicit versions for `Float16` (kinda), `Float32`, and `Float64`, this is basically a fallback.  But I think it is a pretty good one.  (Looks like it's due to @stevengj.)  I did actually do some experimentation with `BigFloat` and `Double64` to see how well it works, and it does pretty darn well.  The only notable thing I found is that we could potentially approximate the number of terms to add [`ceil((0.865*(x1 - 1) / log(x1) - (x1 * 0.0339)) + 1)` is a surprisingly good estimate for sincpi], but I don't think we'd really gain much, and it may not be robust.  The only questionable thing is the choice of 0.5π as the cutoff, and basically I find that it doesn't make much difference.

**Comment by KronosTheLate** [comment]
This solution seems to fix most of the problem in a non-breaking way that makes a lot of sense. I like it.

**Comment by moble** [comment]
The failing tests all seem to be unrelated to this change — things like network errors in the TOML test; missing functions in the cmdlineargs test; something missing from the precompiled sysimg test.  The math test — which I believe is the only thing that actually uses any of the `sinc`/`cosc` functions — passes in every instance I can find.

**Comment by PallHaraldsson** [comment]
Does it help to both have new unnormalized sincu/coscu, and also keyword arguments? Then we will have no consistency in which is used. We could export the former, or not and only have the latter, I'm just not sure people would want to use that way.

**Comment by moble** [comment]
> Does it help to both have new unnormalized sincu/coscu, and also keyword arguments? Then we will have no consistency in which is used.

I'm not devoted to anything in particular, but personally I think that's fine.  The consistency ship has already sailed.  So at this point there's no real cost to having a few extra options around, and it could have the benefit of easing the deprecation path if that is chosen some day.

IMO, the keyword argument does a particularly good job of notifying the user that there's this normalization issue for them to be aware of.  I know I at least frequently read only as far as the signature to see that a method exists, then get back to coding.  So if I see there's a kwarg to consider, I'll stop to read the explanation and think for a second.  That's why I think the kwarg is good.

On the other hand, as someone who does more geometry than signal processing, I'll almost always use the un-normalized version.  So, once I know that another choice exists, I won't want to hassle with the kwarg every time; I'll just use `sincu`/`coscu`.  Similarly, I imagine there are signals people who will want to be explicit about the inclusion of the factor of `pi` — plus the name symmetry is just nice.  That's why I think the non-kwarg options are also good.

**Comment by PallHaraldsson** [comment]
>it could have the benefit of easing the deprecation path if that is chosen some day.

Would it? I think we could deprecate right away, is there any huge cost with using deprecated functions, and people should just migrate to either sincpi or sincu. If we would want to drop sinc, and add it later with a new default, then that would have to be in some very far away future.

Another objection, is what does normalized mean? I would have to think and relalize it means e.g. sincpi.

Arguably you could skip sincu, rather than the keyword. And one reason I could think of for doing that is if we export sincu/coscu/tancu and they would conflict with something. Then it's strictly a breaking change for Julia... and typing Base.sincu or Base.coscu etc. would be rather annoying.

Which do you think is actually more used in practice, the normalized (I guess so, since it was defined) or not?

**Comment by moble** [comment]
> I think we could deprecate right away, is there any huge cost with using deprecated functions?

Hrm.  As a user, I'm not a fan of breaking working code just for nicer names (especially in a non-breaking release).  Don't get me started about, numpy's recent deprecation of `np.product` (which has been around for at least a couple decades) because it's an alias for `np.prod`.  Admittedly, ignoring a warning isn't *technically* breaking, but it's aggravating — especially for those of us a little farther down the OCD spectrum.


> Another objection, is what does normalized mean? I would have to think and relalize it means e.g. sincpi.

Well, it's in the `sinc*` docstrings now, so I'll take your comment that you'd have to think about it as a win for my argument that the kwarg would prompt people to be aware of the issue — which will hopefully lead to more correct math in the world.  And we can all support that!


> Arguably you could skip sincu, rather than the keyword. And one reason I could think of for doing that is if we export sincu/coscu/tancu and they would conflict with something. Then it's strictly a breaking change for Julia... and typing Base.sincu or Base.coscu etc. would be rather annoying.

I'm not sure I understand.  Are you suggesting redefining `sinc` to be what's called `sincu` in this PR?


> Which do you think is actually more used in practice, the normalized (I guess so, since it was defined) or not?

In at least some sense, I would think it's `sincu`, since it comes up so much in geometry (and specifically quaternions), which just gets more cycles than signal processing.

It may also be interesting to note the results of a little survey I conducted while putting this PR together: [Numpy](https://numpy.org/doc/stable/reference/generated/numpy.sinc.html), [Scipy](https://docs.scipy.org/doc/scipy/reference/generated/scipy.special.sinc.html), [Matlab](https://www.mathworks.com/help/signal/ref/sinc.html#d126e196935), and [Maple](https://www.maplesoft.com/support/help/Maple/view.aspx?path=DynamicSystems/Sinc#table40_section229) define `sinc` as $\sin(\pi x)/(\pi x)$; [SymPy](https://docs.sympy.org/latest/modules/functions/elementary.html#sympy.functions.elementary.trigonometric.sinc), [Mathematica](https://reference.wolfram.com/language/ref/Sinc.html), and [R](https://search.r-project.org/CRAN/refmans/rwfec/html/sinc.html) define it as $\sin(x)/(x)$; and [Boost](https://www.boost.org/doc/libs/1_65_1/libs/math/doc/html/math_toolkit/sinc/sinc_pi.html) is the real outlier because it only defines `sinc_pi` which — curiously enough — is $\sin(x)/(x)$.

 
> > no real cost to having a few extra options around
> 
> There's one potential downside. If not inlined, or the part that isn't, e.g. both then sincu and sinc(x, normalized=false) would both take up space in the sysimage? I like it do be smaller, not have two defintions of the same...
> 
> I think any sinc version is actually inlined (or at least part of the implementation), since there's _sincpi, and it inlined.

I admit that I am ill-informed about when things get inlined or not, so I welcome more input about this.

**Comment by stevengj** [comment]
It doesn't make sense to me to have both a `normalized` keyword and `sincu`.

**Comment by stevengj** [comment]
It's also not clear to me why this needs to be in Base, as opposed to in SpecialFunctions.

**Comment by PallHaraldsson** [comment]
>I'm not a fan of breaking working code just for nicer names (especially in a non-breaking release).

I didn't mean that, to me at least deprecation means a signal don't use this in the future. I'm ok with adding just sinpi etc synonyms (and deprecationg older; assuming it doesn't slow down). I believe there's no slowdown (or any indication even, unlike before) running julia on defaults:

--depwarn={yes|no*|error}  Enable or disable syntax and method deprecation warnings (`error` turns warnings into errors)

I mean I'm ok with the rest, unnormalized in SpecialFunctions.jl, but we should discuss the API (here?) either way.

**Comment by PallHaraldsson** [comment]
>ignoring a warning isn't technically breaking, but it's aggravating

Note, there will be no warning, my understanding, unless you opt into, and then silenced by renaming to in your code to sincpi. I believe that's the better (consistent) name, so the rename/synonym fine by me. From it, it follows IMHO that sinc should be deprecated (at least if it doesn't add sincu).

When I think about it, sinc in any form would be best gone, excised, but it can't happen right away, and then deprecation (only) is also logical.

That alone is not going to reduce the sysimage to any noticeable degree, but since you don't think what we have is the more used version, and neither even that common, then it should maybe just be go (in 2.0). cosc in any form is likely even less common, and the logical tanc, isn't here yet, not sure anyone has it or needs it.

What would help (more) in Julia 2.0 would be excising not just sin, sinpi, and all trigonometry and more (also matrix operations in Base, that you have without `using LinearAlgebra`).

I would like new stdlibs, e.g. for this:
```
using Math
using Julia1  # a package that would automatically do the former and more and add all I have planned for excision, for an easy transition
```

>and Boost is the real outlier because it only defines sinc_pi which — curiously enough — is sin(x)/x

Now, that is strange, I thought defining sinpi would signal what is obviously the only sensible meaning...

>Are you suggesting redefining sinc to be what's called sincu in this PR?

No.

**Comment by moble** [comment]
I really like that `Julia1` idea!  Python's `2to3` tool and `from __future__ import ...` mechanism really helped the language to grow; maybe `Julia2` could be the second half of that correspondence.


> cosc in any form is likely even less common

In fact, `cosc(x)/x` is more useful (so perhaps more common) than just `cosc(x)`, and it needs different handling again in the series around 0.  (My Latin is weak, but I suppose that would be the "cosine inferior"  — abbreviated `cosi`?)  Anyway, I'm starting to see the point that all this might be better suited to some other package/stdlib.


> > and Boost is the real outlier because it only defines sinc_pi which — curiously enough — is sin(x)/x
> 
> Now, that is strange, I thought defining sinpi would signal what is obviously the only sensible meaning...

My guess is that the intent is for the normalization of $\mathrm{sinc}_a$ to equal $a$.

**Comment by PallHaraldsson** [comment]
It's good that you like my proposed Julia1 stdlib package. As I see it julia would do `using Julia1` automatically, but the proposed experimental juliax would not:
https://github.com/JuliaLang/julia/pull/51417

I don't see a reason, at least yet, for Julia2 package. What is Julia 2.0, or could be? The Milestone was recently changed to "Potential 2.0" whatever that means, but all changes are in effect deletions and then additions. And additions aren't the problem (or just a question if wanted).

For now I'm thinking only of what could be deleted from Julia, for 2.0 (e.g. LinearAlgebra and OpenBLAS), for lighter compiled packages **by default**. Everyone could still do `using Julia1` with juliax to get that behavior or more specific `using Math` etc.

There are also changes to at least 1 or 2 operators, i.e. to precedence, that are more tricky, since handled by the parser. It would be very breaking to drop |> operator in Julia 2.0, then it has a new behavior in some sense..., I don't know if `using Julia2` would be good for that, I believe a package would actually get JuliaSyntax.jl to changed add operators and keywords.

See also a seemingly important change, that may have to wait for 2.0, if actually breaking (enough...), or just for juliax
https://github.com/JuliaLang/julia/pull/51466
>added breaking and removed minor change labels 3% days ago

File: base/special/trig.jl [hunk 2]
```diff
@@ -1063,31 +1063,86 @@ isinf_real(x::Complex) = isinf(real(x)) && isfinite(imag(x))
...
-    sinc(x)
+    sinc(x; [normalized::Bool=true])
+
+Compute ``\\sin(\\pi x) / (\\pi x)`` if ``x \\neq 0``, and ``1`` if ``x = 0`` (equal to the
+[`sincpi`](@ref) function).  If `normalized` is `false`, the factors of ``\\pi`` are
+omitted (equal to the [`sincu`](@ref) function).
+
+The normalization is with respect to integration over all real values of ``x``:
+```math
+\\int_{-\\infty}^{\\infty} \\mathrm{sincpi}(x) \\, \\mathrm{d}x = 1,
```
Comment (hyrodium) [inline base/special/trig.jl:1074]:
```suggestion
\\int_{-\\infty}^{\\infty} \\operatorname{sincpi}(x) \\, \\operatorname{d}x = 1,
```
It seems `\operatorname` is better here.
https://tex.stackexchange.com/questions/48459/whats-the-difference-between-mathrm-and-operatorname

Comment (moble) [reply base/special/trig.jl:1074]: At least for `sincpi`, I agree that it would be in principle, but I shy away from it because this presumably needs to be rendered with MathJax, which [apparently requires the `ams` package](https://docs.mathjax.org/en/latest/input/tex/macros/index.html#o) to be loaded, and I'm not sure that is likely to be the case wherever the docs would be hosted.  As for `d`, I think it would leave an unwanted space before the `x`.

But in any case, I think this is all moot, since this PR is seemingly destined for the rubbish bin (at least in this repo).

Comment (hyrodium) [reply base/special/trig.jl:1074]: >this presumably needs to be rendered with MathJax, which [apparently requires the ams package](https://docs.mathjax.org/en/latest/input/tex/macros/index.html#o) to be loaded, and I'm not sure that is likely to be the case wherever the docs would be hosted.

It seems KaTeX is used in Documenter.jl, and `\operatorname` is used in other places. See [`eps` docstring](https://github.com/JuliaLang/julia/blob/bed2cd540a11544ed4be381d471bbf590f0b745e/base/float.jl#L1051) and [its rendered documentation](https://docs.julialang.org/en/v1/base/base/#Base.eps-Tuple{AbstractFloat}).

>As for `d`, I think it would leave an unwanted space before the `x`.

Sorry, that was my mistake.

>But in any case, I think this is all moot, since this PR is seemingly destined for the rubbish bin (at least in this repo).

I see. I think adding `sinc` and `sincpi` functions (https://github.com/JuliaLang/julia/issues/36954#issuecomment-1296954308) to SpecialFunctions.jl would be a better solution.

**Comment by PallHaraldsson** [comment]
To not repeat my comment from elsewhere, in short, is it best to (soft) deprecate sinc, and maybe NOT add sincpi (here)? Rather add so SpecialFunctions.jl?

Also is unnormalized sinc, i.e. Sa, simply Sa(x) = pi*sincpi(x/pi)? I mean I think it is, or can be/is defined that way, but regarding accuracy, is it the best implementation?
